### ISSUE TITLE
increase physical address space or available memory for unikernel?

### ISSUE THREAD

**Issue by danbrodsky** [unikraft/unikraft#225]
I have an application I'm attempting to port into a Unikraft project. One issue I'm experiencing is that the physical address space is only 1GB (up to 0x40000000 I believe due to that being the limit for QEMU pc emulator), but my application needs to make allocations that exceed this amount. Using the ptsupport patch from the eurosys21 repository, all `MULTIBOOT_MEMORY_AVAILABLE` regions are used as available physical memory for a page table. Due to how physical memory is laid out [0], the RAM that is allocated to the machine from the hypervisor has a hole around the 3GB mark. The patch does not resize `_libkvmplat_cfg.heap` after the first memory area is mapped in the page table, and so allocators are only initialized with the same amount of memory as the first available multiboot memory area which is ~3GB in size.

I assumed that not resizing `_libkvmplat_cfg.heap` was a mistake in the patch and that if I changed it to the same size as the total amount of available memory in the page table then it would allow my allocator to use all the memory I set in KVM. Instead I end up hitting a page fault when writing to an address just past the 3GB mark, which is where the physical page would have been located in the next section in RAM. What's odd is that the page fault only occurs when the kernel is executing, but when I change the same location in GDB it works just fine. GDB normally cannot read or write to unmapped regions in memory, so it's strange that I can change a value directory that would normally trigger a page fault during execution.

```
[    0.623086] CRIT: [libkvmplat] <traps.c @  123> Page fault at linear address 4fffffffc, rip 5c8c4d, regs 0xd3ff50, sp 7fffff00feb0, our_sp 0xd3ff28, code b
```
(at a breakpoint at 0x5c8c4d)
```
gef➤  set {int} 0x4fffffffc = 1
gef➤  x 0x4fffffffc
0x4fffffffc:	0x00000001
```

I'm assuming that there must be some hardware-level configuration missing in the patch here. Any help is appreciated. It should be easy to reproduce this by using a eurosys21 docker container with the ptsupport patch applied and running a unikernel with  more than 4.5GB of RAM and something like this within the app:
``` C
int main() {
    // allocates 4GB of memory
    for (long unsigned int i = 0; i < 0x400; ++i) {
        printf("allocating memory, total used: 0x%lx\n", i*0x400000);
        void* m = malloc(0x400000);
        if (m == NULL) {
            printf("failed allocation, total: 0x%lx\n", i*0x400000);
            exit(1);
        }
    }
    return 0;
}

```

[0] - https://wiki.osdev.org/Detecting_Memory_(x86)

**Comment by vladandrew** [comment]
Hi Dan,

We use a static page table, entries over `0x000000003fe00000` are not mapped. You should add the mapping for them in `unikraft/plat/kvm/x86/pagetable.S` at line 826. Let me know if it works.

PS: Right, currently we're using only the first available multiboot area, if you have a patch for this, feel free to submit it.

**Comment by danbrodsky** [comment]
Hey Vlad,

> PS: Right, currently we're using only the first available multiboot area, if you have a patch for this, feel free to submit it.

The ptsupport patch already reads the multiboot structure and uses all available memory regions. I edited the patch a bit so it would use the 2nd memory region as `heap2` (wasn't set for some reason). That way it would be mapped and added as memory during allocator initialization. Mapping passes and I can read memory from gdb but it faults during r/w within the allocator. It seems to be dependent on the total amount of memory mapped to the page table so far, as any region that's mapped after over ~3-3.5GB of memory has already been mapped will not be accessible. I tried changing the size and vaddr of both heap regions to see if had an effect, but only the amount of memory mapped changes whether it faults.

**Comment by danbrodsky** [comment]
> We use a static page table, entries over `0x000000003fe00000` are not mapped. You should add the mapping for them in `unikraft/plat/kvm/x86/pagetable.S` at line 826. Let me know if it works.

I tried writing a static page table with 4GB of mapped memory (essentially the same as the one used in rumprun https://github.com/rumpkernel/rumprun/blob/master/platform/hw/arch/amd64/pagetable.S but with more entries). I skipped over the region in memory from 0x80000000 - 0xfffffffc as it doesn't appear to be writeable (MMIO space?). This is my pdpt:
```
cpu_pdpt:
	.quad cpu_pd0 + PAGETABLE_RW
	.quad cpu_pd1 + PAGETABLE_RW
	.quad cpu_pd4 + PAGETABLE_RW
	.quad cpu_pd5 + PAGETABLE_RW
	.quad cpu_pd6 + PAGETABLE_RW
	.fill 0x1fb, 0x8, 0x0
```
Since each entry corresponds to a table that maps 1GB of memory, I skip pd2 and pd3 to avoid the unwriteable region. I also had to make some small changes in setup.c so it would use the larger memory space. This works fine for the buddy allocator, but the TLSF allocator tries to read a very high address during allocation for some reason. I'll try to figure out why later.
